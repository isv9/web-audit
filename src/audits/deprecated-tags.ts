import { AuditResult, WebDocument } from '../web-audit';

const deprecatedTagsNames: string[] = [
  'acronym',
  'big',
  'command',
  'content',
  'dir',
  'element',
  'font',
  'menuitem',
];

export function auditDeprecatedTags(document: Pick<WebDocument, 'getTagAmountMap'>): AuditResult {
  const deprecatedTagAmountMap = document.getTagAmountMap(deprecatedTagsNames);

  const isExistDeprecatedTag = Object.values(deprecatedTagAmountMap).some(
    tagAmount => tagAmount > 0,
  );
  const errors: string[] = [];
  if (isExistDeprecatedTag) {
    errors.push('Document has some deprecated tag');
  }

  return {
    name: 'deprecated elements in dom',
    errors,
    tables: [
      {
        content: deprecatedTagAmountMap,
      },
    ],
  };
}
