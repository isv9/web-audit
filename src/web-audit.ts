import { auditBlockSemantics, auditCommonSemantics, auditHeaderSemantics, auditTextSemantics } from './audits/semantic'
import { auditLinks } from './audits/links'

export type WebDocument = {
  getElementsByTagNameCount (tag: string): number;
  querySelectorAll (query: string): { length: number };
  getEmptyElementsByTagName (tag: string): { length: number }
};

type AuditSectionResult = {
  name: string;
  auditResults: AuditResult[];
};

export type AuditResult = {
  name?: string;
  tables: (AuditResultTable)[];
  liveCollections?: (AuditResultLiveCollection)[];
  logs?: string[];
  warnings?: string[];
  errors?: string[];
};

type AuditResultTable = { name?: string } & { [key: string]: string | number };
export type AuditResultLiveCollection = { name: string, collection: object };

export class WebAudit {

  private readonly webDocument: WebDocument

  constructor (webDocument: WebDocument) {
    this.webDocument = webDocument
  }

  audit () {
    const auditSemanticsResult = this.auditSemantics()
    const auditLinksResult = this.auditLinks()
    console.group('web audit')
    WebAudit.renderAuditSectionResult(auditSemanticsResult)
    WebAudit.renderAuditSectionResult(auditLinksResult)
    console.groupEnd()
  }

  private auditLinks (): AuditSectionResult {
    return {
      name: 'links',
      auditResults: [
        auditLinks(this.webDocument)]
    }
  }

  private auditSemantics (): AuditSectionResult {

    return {
      name: 'semantics',
      auditResults: [auditCommonSemantics(this.webDocument),
        auditHeaderSemantics(this.webDocument),
        auditTextSemantics(this.webDocument),
        auditBlockSemantics(this.webDocument)]
    }
  }

  static renderAuditSectionResult (auditSectionResult: AuditSectionResult) {
    const { name, auditResults } = auditSectionResult
    const details = {
      errorsCount: 0,
      warningsCount: 0
    }
    auditResults.forEach(auditResult => {
      const errorsCount = WebAudit.getAuditResultMessagesCount(auditResult, 'errors')
      const warningsCount = WebAudit.getAuditResultMessagesCount(auditResult, 'warnings')
      details.errorsCount += errorsCount
      details.warningsCount += warningsCount
    })
    console.group(`${name}, (errors="${details.errorsCount}", warnings="${details.warningsCount}")`)
    auditResults.forEach(auditResult => WebAudit.renderAuditResult(auditResult))
    console.groupEnd()
  }

  static renderAuditResult (auditResult: AuditResult) {
    const { name, tables, errors = [], warnings = [], liveCollections = [], logs = [] } = auditResult
    const errorsCount = WebAudit.getAuditResultMessagesCount(auditResult, 'errors')
    const warningsCount = WebAudit.getAuditResultMessagesCount(auditResult, 'warnings')
    console.group(`${name}, (errors="${errorsCount}", warnings="${warningsCount}")`)
    tables.forEach(({ name, nodes, ...tableDataSet }) => {
      console.log(name)
      console.table(tableDataSet)
    })
    liveCollections.forEach(({ name, collection }) => {
      console.log(name)
      console.log(collection)
    })
    logs.forEach(message => console.log(message))
    console.log('Summary')
    if (errorsCount > 0 || warningsCount > 0) {
      errors.forEach(message => console.error(message))
      warnings.forEach(message => console.warn(message))
    } else {
      console.log('It is OK')
    }

    console.groupEnd()
  }

  static getAuditResultMessagesCount (auditResult: AuditResult, key: keyof AuditResult): number {
    let property = auditResult[key]
    if (property === null || property === undefined) {
      property = []
    }
    if (Array.isArray(property)) {
      return property.length
    }
    throw new Error('Property type must be array')
  }
}

