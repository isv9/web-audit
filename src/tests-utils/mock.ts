import { WebDocument } from '../web-audit';

export function createGetTagAmountMapMock(
  getTagAmount: (tag: string) => number,
): WebDocument['getTagAmountMap'] {
  return jest.fn(tagsNames =>
    Object.fromEntries(tagsNames.map(tagName => [tagName, getTagAmount(tagName)])),
  );
}
