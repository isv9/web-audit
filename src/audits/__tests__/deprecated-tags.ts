import { auditDeprecatedTags } from '../deprecated-tags';
import { WebDocument } from '../../web-audit';
import { createGetTagAmountMapMock } from '../../tests-utils/mock';

describe('auditDeprecatedTags', () => {
  it('document does not have any deprecated tags', () => {
    const document: Pick<WebDocument, 'getTagAmountMap'> = {
      getTagAmountMap: createGetTagAmountMapMock(() => 0),
    };
    expect(auditDeprecatedTags(document)).toMatchSnapshot();
  });
  it('document has some deprecated tags', () => {
    const document: Pick<WebDocument, 'getTagAmountMap'> = {
      getTagAmountMap: createGetTagAmountMapMock(() => 1),
    };
    expect(auditDeprecatedTags(document)).toMatchSnapshot();
  });
});
