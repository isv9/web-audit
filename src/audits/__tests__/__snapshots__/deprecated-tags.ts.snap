// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auditDeprecatedTags document does not have any deprecated tags 1`] = `
Object {
  "errors": Array [],
  "name": "deprecated elements in dom",
  "tables": Array [
    Object {
      "content": Object {
        "acronym": 0,
        "big": 0,
        "command": 0,
        "content": 0,
        "dir": 0,
        "element": 0,
        "font": 0,
        "menuitem": 0,
      },
    },
  ],
}
`;

exports[`auditDeprecatedTags document has some deprecated tags 1`] = `
Object {
  "errors": Array [
    "Document has some deprecated tag",
  ],
  "name": "deprecated elements in dom",
  "tables": Array [
    Object {
      "content": Object {
        "acronym": 1,
        "big": 1,
        "command": 1,
        "content": 1,
        "dir": 1,
        "element": 1,
        "font": 1,
        "menuitem": 1,
      },
    },
  ],
}
`;
